Populates backend map from V2 backend map

Weil0ng opened the pull request:
This is a lossy process, we copy everything from v2 map to v1 map but
since v1 map's key type is smaller (16b vs 32b in v2), for entries with
and ID that is larger than 64k, we drop them. This means existing
connection could potentially be interrupted when downgrading from v2 to
v1.

V2 map introduced in https://github.com/cilium/cilium/pull/17235, this is the downgrade path.

Signed-off-by: Weilong Cui <cuiwl@google.com>

Fixes: #17262

pchaigno commented on pkg/service/service.go:
```diff
@@ -219,6 +268,13 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
 	}
 
+	if v2BackendMapExistsV4 || v2BackendMapExistsV6 {
+		log.Info("Backend map v2 exists. Migrating entries with ID < 64k to backend map v1, IDs greater than 64k will be lost.")
```
We could probably count how many >64k entries were lost and emit a warning if we have more than zero. If afraid users will miss this impact if all we log is an info message.

Weil0ng replied:
Done.

pchaigno reviewed (changes requested):
I think this pull request should only be opened against v1.10 given we won't need this code in v1.11.

Weil0ng committed:
Populates backend map from V2 backend map

This is potentially a lossy process, ideally we copy everything from v2
map to v1 map but since v1 map's key type is smaller (16b vs 32b in v2),
for entries with an ID that is larger than 64k, we drop them. This means
existing connection could potentially be interrupted when downgrading
from v2 to v1.

This logic needs to go into 1.10 releases, because we will introduce v2
map in 1.11 and the downgrade logic needs to be in place.

Signed-off-by: Weilong Cui <cuiwl@google.com>

brb commented on pkg/service/service.go:
```diff
@@ -238,6 +305,13 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
 	}
 
+	if v2BackendMapExistsV4 || v2BackendMapExistsV6 {
+		log.Info("Backend map v2 exists. Migrating entries to backend map v1, may lose entries")
```
IMHO "may lose entries" might scare users. The err msg for that case is enough.

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -238,6 +305,13 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
+	if v2BackendMapExistsV4 || v2BackendMapExistsV6 {
+		log.Info("Backend map v2 exists. Migrating entries to backend map v1, may lose entries")
+		if err := s.populateBackendMapV1FromV2(v2BackendMapExistsV4, v2BackendMapExistsV6); err != nil {
+			log.WithError(err).Warn("Error pupolating v1 map from v2 map, might interrupt existing connections during downgrade")
```
Typo: pupolating.

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +189,76 @@ func (s *Service) GetCurrentTs() time.Time {
+		}
+		err := lbmap.Backend4MapV2.DumpWithCallback(copyBackend4Entries)
+		if err != nil {
+			return fmt.Errorf("Unable to populate lb4 backends map: %s", err)
```
s/lb4/IPv4/ and %w for errors.

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +189,76 @@ func (s *Service) GetCurrentTs() time.Time {
+		}
+		err := lbmap.Backend6MapV2.DumpWithCallback(copyBackend6Entries)
+		if err != nil {
+			return fmt.Errorf("Unable to populate lb6 backends map: %s", err)
```
Ditto.

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +189,76 @@ func (s *Service) GetCurrentTs() time.Time {
+			count := 0
+			backendKey := key.(*lbmap.Backend6KeyV2)
+			// See ipv4 above
+			if backendKey.ID > 65535 {
```
Nit: `math.MaxUint16`.

Weil0ng replied:
Done.

brb reviewed (commented):
LGTM, a few nits. Also, I think backporters would appreciate if you could provide more context in the commit msg / PR description why this needs to be backported to v1.10.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +190,76 @@ func (s *Service) GetCurrentTs() time.Time {
+				count = count + 1
+			} else {
+				backend4MapKey := lbmap.NewBackend4Key(backendKey.GetID())
+				lbmap.Backend4Map.Update(backend4MapKey, value.DeepCopyMapValue())
```
We should log an error if updating fails.

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +190,76 @@ func (s *Service) GetCurrentTs() time.Time {
+				backend4MapKey := lbmap.NewBackend4Key(backendKey.GetID())
+				lbmap.Backend4Map.Update(backend4MapKey, value.DeepCopyMapValue())
+			}
+			if count > 0 {
```
This if-statement should be after calling `lbmap.Backend4MapV2.DumpWithCallback()`. Also, `count` should be outside the callback function. Otherwise, for each `ID > max(uint16)` you are going to log the warning msg.

Weil0ng replied:
Good point, moved it outside of the callback func.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +190,76 @@ func (s *Service) GetCurrentTs() time.Time {
 }
 
+func (s *Service) populateBackendMapV1FromV2(ipv4, ipv6 bool) error {
+	removeMap := func(m string) {
```
Have you considered replacing `removeMap` with `(*bpf.Map).Unpin()`?

Weil0ng replied:
Good point, done :)

brb reviewed (changes requested):
Did another round of reviewing. A few things.

brb commented on pkg/service/service.go:
```diff
@@ -187,13 +188,97 @@ func (s *Service) GetCurrentTs() time.Time {
+		err   error
+	)
+
+	if ipv4 {
```
I think we can avoid duplicating the code. A crude way:
```
foo := map[string]bool{"4": ipv4, "6":ipv6}
for k, v := range foo {
if !v {
  continue
}

var map *bpf.Map
var kkey lbmap.BackendKey

if v == "4" {
  map = lbmap.Bakckend4Map
  kkey = key.(*lbmap.Backend4KeyV2)
} else if v == "6" {
 ...
}

rest of the code

}

Weil0ng replied:
Done, PTAL :)

brb reviewed (changes requested):
OK, getting closer to ready :-) Last comment about DRY.
